pkg/apis/pipeline/v1alpha1/task_validation_test.go

```diff
@@ -0,0 +1,162 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"testing"
+
+	buildv1alpha1 "github.com/knative/build/pkg/apis/build/v1alpha1"
+)
+
+var validSource = Source{
+	Name: "source",
+	Type: "git",
+}
+
+func TestTaskSpec_Validate(t *testing.T) {
+	type fields struct {
+		Inputs    *Inputs
+		Outputs   *Outputs
+		BuildSpec *buildv1alpha1.BuildSpec
+	}
+	tests := []struct {
+		name    string
+		fields  fields
+		wantErr bool
+	}{
+		{
+			name:    "nil",
+			wantErr: true,
+		},
+		{
+			name: "valid inputs",
+			fields: fields{
+				Inputs: &Inputs{
+					Sources: []Source{validSource},
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "valid outputs",
+			fields: fields{
+				Outputs: &Outputs{
+					Sources: []Source{validSource},
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "both valid",
+			fields: fields{
+				Inputs: &Inputs{
+					Sources: []Source{validSource},
+				},
+				Outputs: &Outputs{
+					Sources: []Source{validSource},
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "one invalid input",
+			fields: fields{
+				Inputs: &Inputs{
+					Sources: []Source{
+						{
+							Name: "source",
+							Type: "what",
+						},
+						validSource,
+					},
+				},
+				Outputs: &Outputs{
+					Sources: []Source{
+						validSource,
+					},
+				},
+			},
+			wantErr: true,
+		},
+		{
+			name: "one invalid output",
+			fields: fields{
+				Inputs: &Inputs{
+					Sources: []Source{
+						validSource,
+					},
+				},
+				Outputs: &Outputs{
+					Sources: []Source{
+						{
+							Name: "who",
+							Type: "what",
+						},
+						validSource,
+					},
+				},
+			},
+			wantErr: true,
+		},
+		{
+			name: "duplicated inputs",
+			fields: fields{
+				Inputs: &Inputs{
+					Sources: []Source{
+						validSource,
+						validSource,
+					},
+				},
+				Outputs: &Outputs{
+					Sources: []Source{
+						validSource,
+					},
+				},
+			},
+			wantErr: true,
+		},
+		{
+			name: "duplicated outputs",
+			fields: fields{
+				Inputs: &Inputs{
+					Sources: []Source{
+						validSource,
+					},
+				},
+				Outputs: &Outputs{
+					Sources: []Source{
+						validSource,
+						validSource,
+					},
+				},
+			},
+			wantErr: true,
```

Comment:
what do you think about having a separate set of tests for the failure/success conditions vs. having a boolean flag to indicate failure/success?

i find it easier to read tests that are written that way and then the logic in the test itself doesn't have to account for 2 fairly different cases

Reply:
Done.